### ISSUE TITLE
JSON RPC: empty params

### ISSUE THREAD

**Issue by pohly** [spdk/spdk#303]
I am writing a Go client which calls the JSON RPC. I am running into an issue when trying to encode the parameters for  get_bdevs:

```
type GetBDevsArgs struct {
	Name string `json:"name",omitempty`
}
```

Due to the way how Go's encoding/json works, it treats GetBDevsArgs as non-empty and will always emit a (potentially empty) "params" hash. The SPDK server fails to parse that:

```
recvfrom(13, "{\"jsonrpc\":\"2.0\",\"id\":1,\"method\":\"get_bdevs\",\"params\":{}}\n", 32768, 0, NULL, NULL) = 58
write(2, "bdev_rpc.c: 137:spdk_rpc_get_bdevs: *ERROR*: spdk_json_decode_object failed\n", 76) = 76
```

## Expected Behavior

I expect SPDK to accept the empty hash and simply skip over it.

## Current Behavior

`spdk_json_decode_object` fails.

## Remark

Is the `jsonrpc` parameter required by the JSON 2.0 standard? https://godoc.org/github.com/mafredri/cdp/rpcc does not include it, so I had to jump through some hoops to get it included in the stream.

**Comment by dverkamp-intel** [comment]
The JSON-RPC 2.0 spec that SPDK is written against is here: http://www.jsonrpc.org/specification (Note that we don't currently support batch requests, but all other features should be implemented per the spec.)

For the get_bdevs API in particular, the reason it fails when specifying an empty object to params is that the "name" field is not marked optional; the assumption was that if params were specified at all, then "name" would be specified, since it's the only one. This is easy enough to change, and I agree that this makes more sense as an optional parameter: https://review.gerrithub.io/#/c/spdk/spdk/+/411740/

I'm going to search the tree for other RPC methods that have this optional params pattern and fix them as well, but if you find any in particular that aren't working due to this problem, let us know.

As for the `jsonrpc` parameter: by my reading, this is required by a "MUST" in the spec: http://www.jsonrpc.org/specification#request_object - it seems like this is an oversight (or maybe intentional omission) in the Go RPC client you are using.  We could theoretically make this optional, since the only thing we do with the "jsonrpc" value is check that it exactly matches "2.0" and drop the request otherwise, but I'd like to follow what the spec says if possible.

**Comment by dverkamp-intel** [comment]
Actually, the JSON-RPC 2.0 spec also says "Most 2.0 implementations should consider trying to handle 1.0 objects", so I'll take this as confirmation that omitting the `jsonrpc` field is acceptable.  I've posted a patch for review here: https://review.gerrithub.io/#/c/spdk/spdk/+/411742/

**Comment by xuhuagen** [comment]
The error:  spdk_json_decode_object fails can be occur if params is null.
I patch https://review.gerrithub.io/#/c/spdk/spdk/+/411740/ and this issue can be resolved.
or you can not set value  for params filed of your request if params is null and the error will omit.

**Comment by dverkamp-intel** [comment]
The patch that makes the "jsonrpc" request field optional (aa67900a2e42b33386033ed906844d751a57599a) has been merged, as well as the fixes for RPC methods that worked when "params" was omitted entirely but not with an empty "params" (75896c2510ee37821f1760f5090c07d055761b7c and e95f0bb75654ab508522d9a28bc24e0998019491).

@pohly: Does this resolve the issue for you?

If you find any more JSON-RPC interoperability issues, please let us know.

**Comment by wanqunintel** [comment]
@xuhuagen , would you generate a script to automaticly capture this kind of issue for later tracking?

**Comment by pohly** [comment]
Sorry for the late response, I was away last week.

Daniel Verkamp <notifications@github.com> writes:
> The patch that makes the "jsonrpc" request field optional
> (aa67900a2e42b33386033ed906844d751a57599a) has been merged,

That part does not work for me. I found that with SPDK master (rev
9a9bef0a9ae) I still need to send jsonrpc, otherwise I just get back
{\"code\":-32600,\"message\":\"Invalid request\"}} (no further output
from vhost on the console).

But this is okay for me. I am now using a slightly modified version of
net/rpc/json and adding jsonrpc in the modified version is trivial.

> as well as the fixes for RPC methods that worked when "params" was
> omitted entirely but not with an empty "params"
> (75896c2510ee37821f1760f5090c07d055761b7c and
> e95f0bb75654ab508522d9a28bc24e0998019491).

This works, and makes my life easier. I could change the encoding to
suppress "params", but that now is no longer necessary.

> @pohly: Does this resolve the issue for you?

As far as I am concerned, the issue can be closed.

**Comment by dverkamp-intel** [comment]
Sorry, I should have actually tested the optional "jsonrpc" change - it's missing a modification to the decoders to make "jsonrpc" actually optional.  This patch should fix it: https://review.gerrithub.io/#/c/spdk/spdk/+/412859/ (tested manually with telnet this time).

**Comment by wanqunintel** [comment]
@pohly would you verify it with the patch above and update it?

**Comment by pohly** [comment]
I can confirm that it works with that pending patch.

**Comment by dverkamp-intel** [comment]
This should be fixed as of commit da7673f71f15e27a7853ad18f3f07058d73e7274 - thanks for the bug report and tests!
